compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -300,13 +300,9 @@ class Typer extends Namer
                 if (isSelfDenot(defDenot)) curOwner.enclosingClass.thisType
                 else {
                   val effectiveOwner =
-                    if (curOwner.isTerm && defDenot.symbol.isType)
+                    if (curOwner.isTerm && defDenot.symbol.maybeOwner.isType)
```

Comment:
The following alternative check passes all tests too:
```suggestion
                    if (curOwner.isConstructor)
```
This looks more obviously correct to me, wdyt @odersky ?

Reply:
No, I think `curOwner` could also be `localDummy`. `maybeOwner.isType` looks correct.